### ISSUE TITLE
potential bugs for training dynamics 

### ISSUE THREAD

**Issue by hayasick** [FLAIROx/jafar#2]
Hi,

I've been working with the dynamics model and noticed a couple of potential issues. I wanted to check with you to see if my observations are correct:

1. A causal mask is needed for the temporal attention in the ST-Transformer. https://github.com/FLAIROx/jafar/blob/b72f8487584edcfa1741fbd069848327d104879c/utils/nn.py#L50-L54
2. When computing the cross-entropy (CE) loss, the ground truth (GT) labels and the prediction logits need to be shifted. For example:
   ```python
   pred = outputs["token_logits"][:, :-1]
   mask = outputs["mask"][:, :-1]
   target = outputs["video_tokens"][:, 1:]
   ce_loss = optax.softmax_cross_entropy_with_integer_labels(
       pred, target
   )
   ce_loss = (mask * ce_loss).sum() / mask.sum()
   acc = pred.argmax(-1) == target
   ``` 
   https://github.com/FLAIROx/jafar/blob/b72f8487584edcfa1741fbd069848327d104879c/train_dynamics.py#L122-L124

Please let me know if further clarification is needed.

**Comment by EmptyJackson** [comment]
Hey! Thank you for the message and very sorry for the slow response, we're only just getting back to this project :)

We've just pushed a few fixes and sampling code, improving performance greatly. With regards to your questions:
1. Great spot, this slipped out in a refactor. We've now added back the causal mask!
2. Since MaskGIT is trained with an unmasking objective, you don't have to shift the target and predictions. Instead, some of the input sequence tokens are masked, and only the loss on the masked tokens is used to compute the gradient.

Please let me know if you have other questions!
